sympy/stats/stochastic_process_types.py

```diff
@@ -676,31 +771,829 @@ def _transient2absorbing(self):
+        # If we treat p0 the same way as whether it was given or not,
+        # the lower and upper bounds will be wider than necessary for a given p0.
+        # The last example in the docstring would give
+        # Matrix([[AccumBounds(0, 1), AccumBounds(0, 1)]]).
+        # so if p0 is given, we add it into the mix as soon as possible.
+
+        if not p0_is_given:
+            p0 = MatrixSymbol('p_0', 1, n)  # implicitly row-stochastic
+
+        _n = Dummy("n", positive=True, integer=True)
+        if p0_is_given:
+            P_to_the_n = p0 * Matrix(trans_probs) ** _n
+        else:
+            P_to_the_n = Matrix(trans_probs) ** _n
+
+        for row in range(P_to_the_n.shape[0]):
+            for col in range(P_to_the_n.shape[1]):
+                try:  # try to limit pn but it sometimes does not work
+                    P_to_the_n[row, col] = limit_seq(P_to_the_n[row, col], _n)
+                except NotImplementedError:
+                    pass
+
+        if p0_is_given:
+            pn = P_to_the_n
+        else:
+            pn = p0 * P_to_the_n
+        return pn
+
+    def communication_classes(self):
+        """
+        Returns the list of communication classes that partition
+        the states of the markov chain. It will return None if
+        no transition matrix was given.
+
+        Returns
+        =======
+
+        classes : List of CommunicationClass, optional
+            The list of communication classes that make up the
+            Markov Chain.
+
+        Notes
+        =====
+
+        This method uses the following algorithm:
+
+        To find the periods of each state, take the one-step transition
+        matrix P, find P, P^2, ..., P^n where n is the number of states.
+        Analyse the diagonals of these matrices. For state i, find the
+        powers of P for which P^n[i, i] > 0. Find their greatest
+        common divisor of this list for each state.
+
+        To find the communication classes, one could take P^n and find
+        which the states for which j is accessible from i and generate
+        the classes. This, however, loses track of the states the process
+        visited before the nth time step. Instead take X = (P + I)^n and
+        one can build the classes this way. For example, if X[i, j] > 0
+        then j is accessible from i.
+
+        If n is large (about 30), the values in (P + I)^n may also become
+        large (about 2^30). This may cause memory or overflow errors.
+        Replacing I with 0.1*I in the formula could slow this process
+        but it should not be necessary for most uses.
+        """
+        trans_probs = self.transition_probabilities
+        n = self.num_states
+
+        if not self._is_numeric('The transient and recurrent states cannot be determined.'):
+            return None
+
+        temp_Pn = Identity(n)
+        periods = [-1] * n
+        for i in range(1, n + 1):
+            temp_Pn = temp_Pn * trans_probs
```

Comment:
This is computing all powers of `A` up to `A**n` where `A` is an `n x n` matrix. With constant-time arithmetic that's an `O(n**4)` operation which could be very slow for larger matrices. If the matrix has any symbolic entries then the arithmetic might be much worse than constant time.

The strongly connected components function can break the system down in time that is linear in the number of non-zero entries in the matrix (`O(n**2)` for dense matrices but potentially much better for sparse matrices). The kind of calculation you have here could then be done with the resulting smaller matrices.

Reply:
Ok, I agree with that. Given the classes from the strongly connected components function, I've figured out how to identify whether classes are recurrent or transient with something like `O(n**2)` or less. 

I still don't know how to find periodicity quickly, but I think I'll use my same function for the smaller matrices, as you say. This periodicity calculation would still take `O(n**4)` if all states communicate.

Reply:
There is an efficient algorithm for computing the periodicity using breadth-first search here:
http://cecas.clemson.edu/~shierd/Shier/markov.pdf

Reply:
I'll try get this done soon. It's very rare that you find a paper like that which has such a small "acceleration". A lot of the time when I read a paper, if I miss a sentence on the second page, I'm lost for the rest of the paper. Thank goodness that's not the case here. 

Hopefully when this is implemented, I won't even need `CommunicationClass` objects.